File: serve.js
#!/usr/bin/env node

/**
 * Simple HTTP server for local development and testing.
 * Usage: node serve.js [port]
 * Default port: 8000
 */

const http = require('http');
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');

// Default port
const PORT = process.argv[2] || 8000;

// MIME types for common file extensions
const mimeTypes = {
    '.html': 'text/html',
    '.css': 'text/css',
    '.js': 'application/javascript',
    '.json': 'application/json',
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.gif': 'image/gif',
    '.svg': 'image/svg+xml',
    '.ico': 'image/x-icon',
    '.woff': 'font/woff',
    '.woff2': 'font/woff2',
    '.ttf': 'font/ttf',
    '.eot': 'application/vnd.ms-fontobject'
};

const server = http.createServer((req, res) => {
    // Parse URL and remove query parameters
    let filePath = req.url.split('?')[0];
    
    // Default to index.html for root path
    if (filePath === '/') {
        filePath = '/index.html';
    }
    
    // Construct full file path
    const fullPath = path.join(__dirname, filePath);
    
    // Get file extension
    const ext = path.extname(fullPath).toLowerCase();
    const contentType = mimeTypes[ext] || 'text/plain';
    
    // Add CORS headers for local development
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
    
    // Handle OPTIONS requests
    if (req.method === 'OPTIONS') {
        res.writeHead(200);
        res.end();
        return;
    }
    
    // Read and serve the file
    fs.readFile(fullPath, (err, data) => {
        if (err) {
            if (err.code === 'ENOENT') {
                // Try to serve index.html for client-side routing
                if (filePath !== '/index.html') {
                    fs.readFile(path.join(__dirname, 'index.html'), (err, data) => {
                        if (err) {
                            res.writeHead(404, { 'Content-Type': 'text/plain' });
                            res.end('404 Not Found');
                        } else {
                            res.writeHead(200, { 'Content-Type': 'text/html' });
                            res.end(data);
                        }
                    });
                } else {
                    res.writeHead(404, { 'Content-Type': 'text/plain' });
                    res.end('404 Not Found');
                }
            } else {
                res.writeHead(500, { 'Content-Type': 'text/plain' });
                res.end('500 Internal Server Error');
            }
        } else {
            res.writeHead(200, { 'Content-Type': contentType });
            res.end(data);
        }
    });
});

// Start the server
server.listen(PORT, (err) => {
    if (err) {
        console.error('❌ Error starting server:', err);
        process.exit(1);
    }
    
    console.log('🚀 Research Portfolio is running at:');
    console.log(`   http://localhost:${PORT}`);
    console.log(`   http://127.0.0.1:${PORT}`);
    console.log();
    console.log('📱 Test on mobile devices using your local IP:');
    console.log(`   http://[your-local-ip]:${PORT}`);
    console.log();
    console.log('⏹️  Press Ctrl+C to stop the server');
    console.log();
    
    // Try to open browser (cross-platform)
    setTimeout(() => {
        const url = `http://localhost:${PORT}`;
        const start = (process.platform === 'darwin' ? 'open' :
                      process.platform === 'win32' ? 'start' : 'xdg-open');
        exec(`${start} ${url}`, (err) => {
            if (err) {
                console.log(`🌐 Open ${url} in your browser manually`);
            }
        });
    }, 1000);
});

// Handle graceful shutdown
process.on('SIGINT', () => {
    console.log('\n🛑 Server stopped by user');
    process.exit(0);
});

process.on('uncaughtException', (err) => {
    if (err.code === 'EADDRINUSE') {
        console.log(`❌ Port ${PORT} is already in use. Try a different port:`);
        console.log(`   node serve.js ${PORT + 1}`);
    } else {
        console.error('❌ Uncaught Exception:', err);
    }
    process.exit(1);
});

File: serve.py
#!/usr/bin/env python3
"""
Simple HTTP server for local development and testing.
Usage: python serve.py [port]
Default port: 8000
"""

import http.server
import socketserver
import sys
import os
import webbrowser
from threading import Timer

# Default port
PORT = 8000

# Parse command line arguments
if len(sys.argv) > 1:
    try:
        PORT = int(sys.argv[1])
    except ValueError:
        print(f"Invalid port number: {sys.argv[1]}")
        sys.exit(1)

# Change to the directory containing this script
os.chdir(os.path.dirname(os.path.abspath(__file__)))

class MyHTTPRequestHandler(http.server.SimpleHTTPRequestHandler):
    """Custom handler to serve index.html for all paths that don't exist."""
    
    def end_headers(self):
        # Add CORS headers for local development
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        super().end_headers()
    
    def do_GET(self):
        # Serve index.html for root path
        if self.path == '/':
            self.path = '/index.html'
        return super().do_GET()

def open_browser():
    """Open the default web browser after a short delay."""
    webbrowser.open(f'http://localhost:{PORT}')

# Start the server
try:
    with socketserver.TCPServer(("", PORT), MyHTTPRequestHandler) as httpd:
        print(f"🚀 Research Portfolio is running at:")
        print(f"   http://localhost:{PORT}")
        print(f"   http://127.0.0.1:{PORT}")
        print()
        print("📱 Test on mobile devices using your local IP:")
        print(f"   http://[your-local-ip]:{PORT}")
        print()
        print("⏹️  Press Ctrl+C to stop the server")
        print()
        
        # Open browser after 1 second
        Timer(1.0, open_browser).start()
        
        # Start serving
        httpd.serve_forever()
        
except KeyboardInterrupt:
    print("\n🛑 Server stopped by user")
except OSError as e:
    if e.errno == 48:  # Address already in use
        print(f"❌ Port {PORT} is already in use. Try a different port:")
        print(f"   python serve.py {PORT + 1}")
    else:
        print(f"❌ Error starting server: {e}")
    sys.exit(1)
